### PR TITLE
WIP: Adding Image size check

### PR DESCRIPTION
**Pull request by yodahekinsew**
We need to add an image size check to make sure that the promoter isn't promoting unnecessarily large images, since larger images can pose a security risk and are generally bad practice.

### CODE REVIEW

File: cip.go [hunk 1]
```diff
@@ -123,6 +123,10 @@ func main() {
 		"audit-gcp-project-id",
 		os.Getenv("CIP_AUDIT_GCP_PROJECT_ID"),
 		"GCP project ID (name); used for labeling error reporting logs to GCP")
+	maxImageSizePtr := flag.Float64(
+		"max-image-size",
+		2,
```
Comment (listx) [inline cip.go:128]: I would just use MiB instead of GiB. It is more granular with no loss of functionaliy.

File: lib/dockerregistry/inventory.go [hunk 2]
```diff
@@ -1973,6 +1979,42 @@ func (sc *SyncContext) RunChecks(
 	return nil
 }
 
+// Run is a function of ImageSizeCheck and checks that all
+// images to be promoted are under the max file size.
+func (check *ImageSizeCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	byteConvert := 1024.0
```
Comment (listx) [inline lib/dockerregistry/inventory.go:1987]: This variable should be a const, but also, its name doesn't really say much.

File: lib/dockerregistry/inventory_test.go [hunk 3]
```diff
@@ -2049,6 +2049,133 @@ func TestRunChecks(t *testing.T) {
...
+	image2 := reg.Image{
+		ImageName: "bar",
+		Dmap: reg.DigestTags{
+			"sha256:000": {"0.9"}}}
+
+	var tests = []struct {
+		name       string
+		check      reg.ImageSizeCheck
+		manifests  []reg.Manifest
+		imageSizes []int
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2095]: This is a little bit confusing if you have more than 1 image; why not use a map[Digest]int ?

File: lib/dockerregistry/types.go [hunk 4]
```diff
@@ -68,6 +69,14 @@ type PreCheck interface {
 	Run(edges map[PromotionEdge]interface{}) error
 }
 
+// ImageSizeCheck implements the PreCheck interface and checks against
+// images that are larger than a size threshold (controlled by the
+// max-image-size flag).
+type ImageSizeCheck struct {
+	MaxImageSize    int
+	DigestImageSize DigestImageSize
```
Comment (listx) [inline lib/dockerregistry/types.go:77]: We already store this information inside SyncContext... why do we need it here as well?

File: lib/dockerregistry/inventory.go [hunk 5]
```diff
@@ -1276,6 +1277,11 @@ func (sc *SyncContext) ReadRegistries(
 					fmt.Printf("digest %s: %s\n", digest, err)
 				}
 				sc.DigestMediaType[Digest(digest)] = mediaType
+
+				// Store ImageSize
+				if sc.DigestImageSize != nil {
```
[the remark this answers is not in the document]
Comment (yodahekinsew) [reply lib/dockerregistry/inventory.go:1282]: Because some tests make a SyncContext without the DigestImageSize field which causes the line "sc.DigestImageSize[Digest(digest)] = int(mfestInfo.Size)" to fail

File: lib/dockerregistry/types.go (hunk 4 above)
Comment (yodahekinsew) [reply lib/dockerregistry/types.go:77]: because the sync context isn't accessible within the check.Run() function. would it be better to just pass in the SyncContext itself?

File: lib/dockerregistry/inventory_test.go [hunk 6]
```diff
@@ -2049,6 +2049,133 @@ func TestRunChecks(t *testing.T) {
...
+		manifests []reg.Manifest
+		expected  error
+	}{}
+
+	for _, test := range tests {
+		edges, _ := reg.ToPromotionEdges(test.manifests)
+		got := test.check.Run(edges)
+		err := checkEqual(got, test.expected)
+		checkError(t, err,
+			fmt.Sprintf("checkError: test: %v (Error Tracking)\n", test.name))
```
[the remark this answers is not in the document]
Comment (yodahekinsew) [reply lib/dockerregistry/inventory_test.go:2065]: Yeah it is

File: lib/dockerregistry/inventory_test.go [hunk 7]
```diff
@@ -2042,13 +2039,123 @@ func TestRunChecks(t *testing.T) {
...
+			[]reg.Manifest{
+				{
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": .5 * 1024 * 1024,
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2094]: Use a helper function instead of 1024 * 1024.

Or, use a constant and multiply by it (toMB).

File: lib/dockerregistry/inventory_test.go [hunk 8]
```diff
@@ -2042,13 +2039,123 @@ func TestRunChecks(t *testing.T) {
...
 	for _, test := range tests {
-		got := sc.RunChecks(edges, test.checks)
+		got := sc.RunChecks(test.checks)
 		err := checkEqual(got, test.expected)
 		checkError(t, err,
 			fmt.Sprintf("checkError: test: %v (Error Tracking)\n", test.name))
 	}
 }
 
+func TestImageSizeCheck(t *testing.T) {
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2049]: You should also add cases for negative image size, what happens when you pass in -max-image-size=0, etc.

Always consider as many edge cases as possible...!

File: lib/dockerregistry/inventory.go [hunk 9]
```diff
@@ -1973,6 +1976,60 @@ func (sc *SyncContext) RunChecks(
...
+	if len(oversizedImages) > 0 {
+		sort.Strings(oversizedImages)
+		errStr += fmt.Sprintf("The following images were over the max file "+
+			"size of %dMiB: %v",
+			check.MaxImageSize,
+			strings.Join(oversizedImages, ", "))
+	}
+	if len(invalidImages) > 0 {
+		sort.Strings(invalidImages)
+		errStr += fmt.Sprintf("The following images had an invalid file size "+
```
Comment (listx) [inline lib/dockerregistry/inventory.go:2008]: You should just display the actual image file size found for the particular image.

File: lib/dockerregistry/inventory.go [hunk 10]
```diff
@@ -1973,6 +1976,60 @@ func (sc *SyncContext) RunChecks(
...
+			oversizedImages = append(oversizedImages,
+				string(edge.DstImageTag.ImageName))
+		}
+		if imageSize <= 0 {
+			invalidImages = append(invalidImages,
+				string(edge.DstImageTag.ImageName))
+		}
+	}
+
+	errStr := ""
```
Comment (listx) [inline lib/dockerregistry/inventory.go:1998]: You can just use fmt.Errorf() directly and return it when you use it, instead of appending into this errStr.

File: lib/dockerregistry/inventory.go [hunk 11]
```diff
@@ -1973,6 +1976,60 @@ func (sc *SyncContext) RunChecks(
...
+	if len(oversizedImages) > 0 {
+		sort.Strings(oversizedImages)
+		errStr += fmt.Sprintf("The following images were over the max file "+
```
Comment (listx) [inline lib/dockerregistry/inventory.go:2001]: Better to display the actual image size for the offending image (probably in bytes but also in MiB).

File: lib/dockerregistry/inventory.go [hunk 12]
```diff
@@ -1973,6 +1976,60 @@ func (sc *SyncContext) RunChecks(
...
+			check.MaxImageSize,
+			strings.Join(oversizedImages, ", "))
+	}
+	if len(invalidImages) > 0 {
+		sort.Strings(invalidImages)
+		errStr += fmt.Sprintf("The following images had an invalid file size "+
+			"of 0 bytes or less: %v",
+			strings.Join(invalidImages, ", "))
+	}
+	if errStr != "" {
```
Comment (listx) [inline lib/dockerregistry/inventory.go:2012]: This branch is not necessary if you don't use errStr.

File: lib/dockerregistry/inventory_test.go [hunk 13]
```diff
@@ -2042,13 +2039,187 @@ func TestRunChecks(t *testing.T) {
...
+	image1 := reg.Image{
+		ImageName: "foo",
+		Dmap: reg.DigestTags{
+			"sha256:000": {"0.9"}}}
+	image2 := reg.Image{
+		ImageName: "bar",
+		Dmap: reg.DigestTags{
+			"sha256:111": {"0.9"}}}
+
+	const mbToBytes = 1024 * 1024
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2072]: Another way to do this is to do a bitwise left-shift. But this is fine.

File: lib/dockerregistry/inventory_test.go [hunk 14]
```diff
@@ -2042,13 +2039,187 @@ func TestRunChecks(t *testing.T) {
...
+			},
+			map[reg.Digest]int{
+				"sha256:000": 5 * mbToBytes,
+				"sha256:111": 10 * mbToBytes,
+			},
+			fmt.Errorf("The following images were over the max file size" +
+				" of 1MiB: bar, foo"),
+		},
+		{
+			"Max image size is zero",
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2143]: I don't think it's reasonable to allow a max image size of 0... This is a case where we have to guard against misbehaving humans who put in nonsense values.

File: lib/dockerregistry/inventory_test.go [hunk 15]
```diff
@@ -2042,13 +2039,187 @@ func TestRunChecks(t *testing.T) {
...
+			map[reg.Digest]int{
+				"sha256:000": 5 * mbToBytes,
+			},
+			fmt.Errorf("The following images were over the max file size" +
+				" of 0MiB: foo"),
+		},
+		{
+			"Max image size is negative",
+			reg.ImageSizeCheck{
+				MaxImageSize:    -10,
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2165]: Same here, we should probably guard our image size flag value so that only positive integers are accepted. Then we don't have to test it down the line here during the check itself.

File: lib/dockerregistry/inventory_test.go [hunk 16]
```diff
@@ -2042,13 +2039,187 @@ func TestRunChecks(t *testing.T) {
...
+					Registries: registries,
+					Images: []reg.Image{
+						image1,
+						image2,
+					},
+					SrcRegistry: &srcRC},
+			},
+			map[reg.Digest]int{
+				"sha256:000": 0,
+				"sha256:111": -5 * mbToBytes,
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:2199]: Great, this is what I was thinking of as edge cases.

File: lib/dockerregistry/inventory.go [hunk 17]
```diff
@@ -1953,13 +1957,12 @@ func MKPopulateRequestsForPromotionEdges(
 
 // RunChecks runs defined PreChecks in order to check the promotion.
 func (sc *SyncContext) RunChecks(
-	edges map[PromotionEdge]interface{},
```
Comment (listx) [inline lib/dockerregistry/inventory.go:1956]: I didn't check directly, but can you make sure that these diffs that are not related to your image size check are in a separate commit?

File: lib/dockerregistry/inventory_test.go [hunk 18]
```diff
@@ -1993,23 +1995,18 @@ func TestCheckOverlappingEdges(t *testing.T) {
 
 type FakeCheckAlwaysSucceed struct{}
 
-func (c *FakeCheckAlwaysSucceed) Run(
```
Comment (listx) [inline lib/dockerregistry/inventory_test.go:1996]: This stuff around edges should also be in a separate commit (I didn't check, can you verify?). I wonder if you rebased too far back into your history?

File: lib/dockerregistry/inventory.go (hunk 10 above)
Comment (yodahekinsew) [reply lib/dockerregistry/inventory.go:2017]: I'm using an errStr so that I can bubble up errors for both oversized images and invalid images. if I use fmt.Errorf() directly how would I bubble up the two errors into one

File: lib/dockerregistry/inventory.go (hunk 17 above)
Comment (listx) [reply lib/dockerregistry/inventory.go:1956]: Ping?

File: lib/dockerregistry/inventory_test.go (hunk 18 above)
Comment (listx) [reply lib/dockerregistry/inventory_test.go:1996]: Ping? This change is also in your other PR for image removal. You should only have it in 1 of them.

**Comment by yodahekinsew** [comment]
Moved to PR #233
